### DISCUSSION TITLE
Implementing getter for vector of class instances (without copy)

### DISCUSSION THREAD

**Discussion by chrisjsewell** [PyO3/pyo3#3223] [Questions] [answered]
Heya, given the following class, how can I expose the `children` attribute in Python?

```rust
#[pyclass]
pub struct Node {
    pub type_: String,
    pub children_: Vec<Node>,
}
```

I know I can do the following, but then it is always creating new copies of the nodes and I want just the references:

```rust
#[pyclass]
#[derive(Clone)]
pub struct Node {
    pub type_: String,
    pub children_: Vec<Node>,
}

#[pymethods]
impl Node {
    #[getter]
    fn children(&self) -> Vec<Node> {
        self.children_.clone()
    }
}
```

---

For bonus points, I would also like to implement a `walk` method to iterate over `Node` references in the tree

**Comment by adamreichold** [comment]
It depends, if you want to share access to the nodes with Python code, you probably to store it in `Py<Node>`. `Vec<Py<Node>` would still create a new vector, but it will not clone the nodes, just increase their reference counts. (This is basically also what native Python containers like `PyList` do.)

Additionally and orthogonally to that, you could turn `Node` itself into a Python container by implementing slots like `__getitem__` etc. and have them return `Py<Node>`, but that is more a question of API ergonomics than where state is shared or not.

**Reply by chrisjsewell** [reply]
Thanks @adamreichold could you provide me an example of how this would work in the code above and how you "create" `Py<Node>`?

**Reply by adamreichold** [reply, marked as answer]
```rust

#[pyclass]
#[derive(Clone)]
pub struct Node {
    pub type_: String,
    pub children_: Vec<Py<Node>>,
}

#[pymethods]
impl Node {
    #[getter]
    fn children(&self) -> Vec<Py<Node>> {
        self.children_.clone()
    }

    fn add_child(&mut self, type_: String, py: Python<'_>) {
        let child = Self {
            type_,
            children_: Vec::new(),
        };

        self.children_.push(Py::new(py, child).unwrap());
    }
}
```

**Reply by chrisjsewell** [reply]
Awesome, thanks! that seems to work 😄 

Just to note, unless I'm missing something I think with this implementation the `Clone` and separate getter are no longer required, i.e.

```rust
#[pyclass]
pub struct Node {
    pub type_: String,

	#[pyo3(get)]
    pub children: Vec<Py<Node>>,
}

#[pymethods]
impl Node {
    fn add_child(&mut self, type_: String, py: Python<'_>) {
        let child = Self {
            type_,
            children: Vec::new(),
        };

        self.children.push(Py::new(py, child).unwrap());
    }
}
```

**Reply by chrisjsewell** [reply]
@adamreichold another quick question 😅 

how can you return a reference of `self` from a method?

```rust
#[pyclass]
pub struct MyClass {}

#[pymethods]
impl MyClass {
	fn change_state(&mut self) -> ?? {
        // change state ...
        return self
    }
}
```

e.g. to chain method calls together on the python side

```python
my_class = MyClass().change_state().change_state()
```

**Reply by adamreichold** [reply]
Instead of `&self`, take `slf: Py<Self>` as the receiver and return `Py<Self>` as well. (Use `slf.borrow(py)` to get access to `&self` in that mode.)

**Reply by chrisjsewell** [reply]
thats the one, thanks again!

**Reply by krai33** [reply]
As an extand question for myself, my struct `Node` comes from an external lib.rs with is not defined with `#[pyclass]`. Code above `Py::new(py, child)` could not be done due to `Node` the trait `PyClass` is not implemented. What should i do to avoid the data copy?
My rust struct is defined like:
``` rust
pub struct Node{
    pub value_a: u32,
    pub value_b: Vec<f32>,
}

**Reply by aander80** [reply]
@krai33 Did you solve this issue to avoid copying the data?
